src/overlay/OverlayManagerImpl.cpp

```diff
@@ -301,53 +308,92 @@ OverlayManagerImpl::getPeersList(Peer* peer)
 }
 
 void
-OverlayManagerImpl::storePeerList(std::vector<std::string> const& list,
-                                  bool setPreferred)
+OverlayManagerImpl::storePeerList(std::vector<PeerBareAddress> addresses,
+                                  bool setPreferred, bool startup)
 {
     auto typeUpgrade = setPreferred
                            ? PeerManager::TypeUpdate::SET_PREFERRED
                            : PeerManager::TypeUpdate::UPDATE_TO_OUTBOUND;
-
-    for (auto const& peerStr : list)
+    for (auto const& peer : addresses)
     {
-        try
+        if (setPreferred)
+        {
+            mConfigurationPreferredPeers.insert(peer);
+        }
+
+        if (startup)
         {
-            auto address = PeerBareAddress::resolve(peerStr, mApp);
-            getPeerManager().update(address, typeUpgrade,
+            getPeerManager().update(peer, typeUpgrade,
                                     PeerManager::BackOffUpdate::HARD_RESET);
         }
-        catch (std::runtime_error&)
+        else
         {
-            CLOG(ERROR, "Overlay") << "Unable to add peer '" << peerStr << "'";
+            // If address is present in the DB, `update` will ensure
+            // type is correctly updated. Otherwise, a new entry is created.
+            // Note that this won't downgrade promoted peers back to outbound.
+            getPeerManager().update(peer, typeUpgrade);
         }
     }
 }
 
 void
-OverlayManagerImpl::storeConfigPeers()
+OverlayManagerImpl::storeKnownAndPreferred(
+    std::vector<PeerBareAddress> known, std::vector<PeerBareAddress> preferred,
+    bool startup)
+{
+    storePeerList(std::move(known), false, startup);
+    storePeerList(std::move(preferred), true, startup);
+};
+
+void
+OverlayManagerImpl::storeConfigPeers(bool startup)
+{
+    if (startup)
+    {
+        // On startup, perform IP resolution synchronously on the main thread
+        auto known = resolvePeers(mApp.getConfig().KNOWN_PEERS);
+        auto preferred = resolvePeers(mApp.getConfig().PREFERRED_PEERS);
+        storeKnownAndPreferred(known, preferred, startup);
+    }
+    else
+    {
+        mApp.postOnBackgroundThread(
+            [this, startup]() {
+                auto known = resolvePeers(this->mApp.getConfig().KNOWN_PEERS);
```

Comment:
as we're on a different thread...
1. how do we guarantee that `this` is valid? 
2. how do we ensure that things don't blow up on shutdown? In particular, is it safe to have this run for a long time while the main thread is tearing down overlay?

Reply:
looks to me that a pattern that uses `std::future` would be cleaner and easier to reason about, you could `peek` into this future from the `tick` cb

Reply:
Agreed that `std::future` seems to be safer in this case. 
To address your questions:
1. From what I see, when application is destructing, it first joins all threads (while blocking), before destroying its member variables. This means that the background thread will finish its work while the app (hence `mOverlayManager`, and hence captured `this`) is still alive. 
2. I think the right course of action here would be to stop any activity as soon as OverlayManager starts shutting down, i.e. stop ticking and resolving any peers.